numpy/ma/core.py

```diff
@@ -4984,23 +4877,51 @@ def mean(self, axis=None, dtype=None, out=None):
 
         """
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = super(MaskedArray, self).mean(axis, dtype, out, **kwargs)
+            zerocounts = self.count(axis, **kwargs) == 0
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    if any(zerocounts):
+                        out.__setmask__(zerocounts)
+                    else:
+                        out.__setmask__(nomask)
+                return out
+            # mask in case of empty slices
+            if any(zerocounts):
+                ret = ret.view(MaskedArray)
+                ret.__setmask__(zerocounts)
+            return ret
+
+        # code largely copied from _methods.mean
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
+        if ret is masked:
+            return _return_masked(out)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = np.true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
```

Comment:
Also, I just checked removing `subok=False` in the calls in `_methods.py` does not cause any unit tests to fail. So instead of blindly copying from there, it seems we might just remove subok both here and in that file.